File: src/main/scala/com/getouo/gb/scl/rtp/H264FileReader.scala
package com.getouo.gb.scl.rtp

import java.io.FileInputStream
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.atomic.AtomicReference

class H264FileReader(fileName: String, consumer: H264NALU => Unit) extends Runnable {

//  private var fis = new FileInputStream(fileName)
  private var fis: FileInputStream = null
  private val startContainer: AtomicReference[Array[Byte]] = new AtomicReference[Array[Byte]](null)
//  private var consumer: H264NALU => Unit = null

  override def run(): Unit = {
    while (true) {
      fis = new FileInputStream(fileName)
      var splitBuf = new Array[Byte](0)
      val buf = new Array[Byte](1024 * 8)
      var readLen = fis.read(buf)
      while (readLen != -1) {
        splitBuf ++= buf.slice(0, readLen)
        var tag4Index = H264FileAccessor.startTag4Index(splitBuf)
        var tag3Index = H264FileAccessor.startTag3Index(splitBuf)
        while (tag4Index != -1 || tag3Index != -1) {
          val lastStart: Array[Byte] = startContainer.get()
          if (tag4Index != -1) {
            if (lastStart != null) pop(H264NALU(lastStart.length, splitBuf.slice(0, tag4Index)))
            splitBuf = splitBuf.drop(tag4Index + 4)
            startContainer.set(H264FileAccessor.START_TAG4)
          } else if (tag4Index == -1 && tag3Index != -1) {
            if (lastStart != null) pop(H264NALU(lastStart.length, splitBuf.slice(0, tag3Index)))
            splitBuf = splitBuf.drop(tag3Index + 3)
            startContainer.set(H264FileAccessor.START_TAG3)
          }
          tag4Index = H264FileAccessor.startTag4Index(splitBuf)
          tag3Index = H264FileAccessor.startTag3Index(splitBuf)
        }
        readLen = fis.read(buf)
      }
      fis.close()
      val last = startContainer.get()
      if (last != null && !splitBuf.isEmpty) pop(H264NALU(last.length, splitBuf))
      startContainer.set(null)
      System.err.println("file read finished!")
    }
  }

  private def pop(n: H264NALU): Unit = {
    try {
      if (consumer != null)
        consumer.apply(n)
//      ioQueue.add(n)
    } catch {
      case e: Throwable =>
    }
  }
}

File: src/main/java/com/getouo/gb/util/H264FileReader.java
package com.getouo.gb.util;

import com.getouo.gb.scl.rtp.RtpPacket;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.concurrent.ArrayBlockingQueue;

public class H264FileReader implements Runnable {

    private final FileInputStream fis;
    private final ArrayBlockingQueue<RtpPacket> packetsQueue;
    final ByteBuffer buffer = ByteBuffer.allocateDirect(1024 * 32);

    public H264FileReader(String fileName, ArrayBlockingQueue<RtpPacket> packetsQueue) throws FileNotFoundException {
        this.fis = new FileInputStream(fileName);
        this.packetsQueue = packetsQueue;
    }


    @Override
    public void run() {
        while (true) {
            try {
                boolean bufferIsEmpty = false;
                int readLen = -1;
                byte[] buf = new byte[1024 * 8];
                while (this.fis.read(buf) != -1) {
                    if (bufferIsEmpty) {
                        for (int i = 0; i < readLen; i++) {
                            if (i + 3 < readLen && buf[i] == 0 && buf[i+1] == 0 && buf[i+2] == 0 && buf[i+3] == 1) { // start4
                                buffer.put(new byte[]{buf[i], buf[i + 1], buf[i + 2], buf[i + 3]});
                                i += 3;
                                bufferIsEmpty = true;
                            } else if (i + 2 < readLen && buf[i] == 0 && buf[i+1] == 0 && buf[i+2] == 1) {
                                buffer.put(new byte[]{buf[i], buf[i + 1], buf[i + 2]});
                                i += 2;
                                bufferIsEmpty = true;
                            }
                        }
                    } else {

                    }

                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}
